SCRUM-14 forgot password and email OTP feature added to BE, error han…

alcarrillomunoz opened the pull request:
…dling improved

Description
This merge addresses the forgot password feature for the backend, adding OTP email feature, and fixes error handling issues with improved error messaging for API calls. 

Changes Made
installed express-async-errors package to help with async errors crashing program, added some error-handling middleware to send error messages for API calls (front end can access with 'data.msg' variable) created forgot password routes/controllers, installed nodemailer, set email variables in .env file (will need to adjust per team decision - currently using a temporary email service simulation program via Ethereal), configured reset password controller to generate OTP (expires after 10 minutes, sent via email to user, user enters code and if valid is granted access to reset password, added isModified function to check for modifications of password and trigger 'save' function to re-hash new passwords in db

Related Issues
Will need to change email settings depending on what team decides, and can add styling to email to match project style. 

Additional Notes
Testing was done via Postman and Ethereal email simulation program. Functions were testing correctly. .env file with applicable sensitive data is saved locally on my computer.  

Merge Request Checklists
 Code follows project coding guidelines.
 Documentation reflects the changes made.

Miminiverse commented on src/utils/sendEmail.js:
emailOTP is an async function but there's no await. I'd recommend to add await in transporter.sendMail with a try catch

Miminiverse commented on src/controllers/resetPassword.js:
I's say this method of generating random number is simple to implement and works well for our current app. But there are some weaknesses, for example numbers like 042813 become 42813 (5 digits) when stored as a number.
In production, I'd recommend to use some library to enhance security like crypto library.

Miminiverse replied:
There are some security concerns I would like to mention here: 
Race Conditions: if multiple users request OTPs simultaneously, globalThis.newOTP will be overwritten, invalidating previous OTPs. All users share the same global OTP. User A could use User B's OTP.
Persistence: OTPs are lost if the server restarts/crashes. There's no way we can invalidate the OTP. 
Return: it's correct to send OTP via email only, and we should not send the newOTP to client in res.status(StatusCodes.OK).json
There are some solutions to these issues and a common approach is to store the OTP in the database associate with each user. 
However since we do not have a lot of time time to, I'd recommend to consider reset password  as a nice-to-have feature (not essential for now) and we can keep this approach, just let me know your acknowledgment of these issues and I can approve this PR. 
Please also remove console.log("OTP expired");

Miminiverse commented on src/models/User.js:
From line 32 to 26 is quite redundant, is there any reasons you handle it this way? 
Would you consider this way?
      const salt = await bcrypt.genSalt(10);
      this.password = await bcrypt.hash(this.password, salt);

alcarrillomunoz commented:
@Miminiverse thanks so much for the feedback! I made the changes requested: 
- added await in transporter.sendMail with a try catch
- removed console.log("OTP expired");
- removed redundancy in password hashing, per suggested form 
- moved next() to try block 

About security concerns, that absolutely makes sense! I'll consider all of this moving forward, and will look into libraries with security concerns in mind. I've submitted a new commit for review with edits listed above.

Miminiverse reviewed (approved):
Such an effort in this Lily! 
Nice implementation to hash the password whenever we need to update it
